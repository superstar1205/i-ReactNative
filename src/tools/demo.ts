export enum CommentType {
  REMOVE_FILE = `@demo remove-file`,
  REMOVE_CURRENT_LINE = `@demo remove-current-line`,
  REMOVE_NEXT_LINE = `@demo remove-next-line`,
  REMOVE_BLOCK_START = `@demo remove-block-start`,
  REMOVE_BLOCK_END = `@demo remove-block-end`,
}

/**
 * Take the file content as a string and remove any
 * line of code with an `// @demo remove-current-line` comment
 */
function removeCurrentLine(contents: string, comment = CommentType.REMOVE_CURRENT_LINE): string {
  const lines = contents.split("\n")
  const result = lines.filter((line) => !line.includes(comment))
  return result.join("\n")
}

/**
 * Take the file content as a string and remove the next line
 * of code with an `// @demo remove-next-line` comment before it
 */
function removeNextLine(contents: string, comment = CommentType.REMOVE_NEXT_LINE): string {
  const lines = contents.split("\n")
  const result = lines.filter((line, index) => {
    const prevLine = lines[index - 1]

    const preserveCurrent = line.includes(comment) === false
    const preservePrevious = prevLine !== undefined && prevLine.includes(comment) === false

    if (index === 0) {
      // if we are on the first line, there is no previous line to check
      return preserveCurrent
    }

    // keep current line if there is no comment in current or previous line
    const keepLine = preserveCurrent && preservePrevious
    return keepLine
  })
  return result.join("\n")
}

/**
 * Take the file content as a string and remove the lines of code between
 * `// @demo remove-block-start` and `// @demo remove-block-end` comments
 */
function removeBlock(
  contents: string,
  comment = { start: CommentType.REMOVE_BLOCK_START, end: CommentType.REMOVE_BLOCK_END },
): string {
  const { start, end } = comment
  const lines = contents.split("\n")

  const findIndex = (l: typeof lines, c: typeof start | typeof end) =>
    l.findIndex((line) => line.includes(c))
  const NOT_FOUND = -1

  const blockStartIndex = findIndex(lines, start)
  const blockEndIndex = findIndex(lines, end)
  const blockExists = findIndex(lines, start) !== NOT_FOUND && blockEndIndex !== NOT_FOUND

  if (blockExists) {
    const blockLength = blockEndIndex - blockStartIndex + 1
    lines.splice(blockStartIndex, blockLength) // mutates `lines`
  }

  const updateContents = lines.join("\n")

  const anotherBlockExists =
    findIndex(lines, start) !== NOT_FOUND && findIndex(lines, end) !== NOT_FOUND
  if (anotherBlockExists) {
    return removeBlock(updateContents, comment)
  }

  return updateContents
}

/**
 * Perform all remove operations possible in a file
 * @param contents The file contents as a string
 * @return The file contents with all remove operations performed
 */
function remove(contents: string): string {
  const result = removeBlock(removeNextLine(removeCurrentLine(contents)))
  return result
}

export const demo = {
  CommentType,
  removeCurrentLine,
  removeNextLine,
  removeBlock,
  remove,
} as const
