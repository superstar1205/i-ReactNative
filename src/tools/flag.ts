/**
 * Utility for converting 'true' and 'false' strings to booleans.
 *
 * Both `Boolean('true')` and `Boolean('false')` return `true`, because any string is a truthy value.
 * `!!'true'` and `!!'false'` also return `true`, for the same reason.
 * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean
 *
 * @param value The value to check.
 * @returns `true` for 'true', `false` for 'false', `Boolean` result otherwise.
 */
export function bool(value: unknown): boolean {
  if (value === "false") return false
  return Boolean(value)
}

/**
 * Utility for converting 'true' and 'false' strings to booleans,
 * while preserving `undefined` values as `undefined` instead of `false`.
 */
export function boolFlag(option: unknown): boolean | undefined {
  if (option === undefined) return undefined
  return bool(option)
}
