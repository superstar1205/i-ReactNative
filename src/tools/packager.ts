import { system } from "gluegun"
import { spawnProgress } from "./spawn"

// we really need a packager core extension on Gluegun
// in the meantime, we'll use this hacked together version

type PackageInstallOptions = {
  dev?: boolean
  expo?: boolean
  onProgress?: (out: string) => void
}
const packageInstallOptions: PackageInstallOptions = {
  dev: false,
  expo: false,
  onProgress: (out: string) => console.log(out),
}

type PackageListOptions = {
  packager?: "npm" | "yarn"
  global?: boolean
}
const packageListOptions: PackageListOptions = {
  global: false,
}

let isYarn
function yarn() {
  if (isYarn !== undefined) return isYarn
  isYarn = Boolean(system.which("yarn"))
  return isYarn
}

function add(pkg: string, options: PackageInstallOptions = packageInstallOptions) {
  if (options.expo) {
    return `npx expo-cli install ${pkg}`
  } else if (yarn()) {
    const dev = options.dev ? " --dev" : ""
    return `yarn add ${pkg}${dev}`
  } else {
    const dev = options.dev ? " --save-dev" : ""
    return `npm install ${pkg}${dev}`
  }
}

function remove(pkg: string, options: PackageInstallOptions = packageInstallOptions) {
  if (options.expo) {
    return `npx expo-cli uninstall ${pkg}`
  } else if (yarn()) {
    return `yarn remove ${pkg}`
  } else {
    return `npm uninstall ${pkg}`
  }
}

function install() {
  if (yarn()) {
    return `yarn install -s`
  } else {
    return `npm install`
  }
}

type PackageListOutput = [string, (string) => [string, string][]]
function list(options: PackageListOptions = packageListOptions): PackageListOutput {
  if (options.packager === "yarn" || (options.packager === undefined && yarn())) {
    return [
      `yarn${options.global ? " global" : ""} list`,
      (output: string): [string, string][] => {
        // Parse yarn's human-readable output
        return output.split("\n").reduce((acc: [string, string][], line: string): [
          string,
          string,
        ][] => {
          const match = line.match(/info "([^@]+)@([^"]+)" has binaries/)
          return match ? [...acc, [match[1], match[2]]] : acc
        }, [])
      },
    ]
  } else {
    return [
      `npm list${options.global ? " --global" : ""} --depth=0 --json`,
      (output: string): [string, string][] => {
        // npm returns a single JSON blob with a "dependencies" key
        const json = JSON.parse(output)
        return Object.keys(json.dependencies || []).map((key: string): [string, string] => [
          key,
          json.dependencies[key].version,
        ])
      },
    ]
  }
}

export const packager = {
  run: async (command: string, options: PackageInstallOptions = packageInstallOptions) => {
    if (yarn()) {
      return spawnProgress(`yarn ${command}`, { onProgress: options.onProgress })
    } else {
      return spawnProgress(`npm run ${command}`, { onProgress: options.onProgress })
    }
  },
  add: async (pkg: string, options: PackageInstallOptions = packageInstallOptions) => {
    const cmd = add(pkg, options)
    return spawnProgress(cmd, { onProgress: options.onProgress })
  },
  remove: async (pkg: string, options: PackageInstallOptions = packageInstallOptions) => {
    const cmd = remove(pkg, options)
    return spawnProgress(cmd, { onProgress: options.onProgress })
  },
  install: async (options: PackageInstallOptions = packageInstallOptions) => {
    const cmd = install()
    return spawnProgress(cmd, { onProgress: options.onProgress })
  },
  list: async (options: PackageListOptions = packageListOptions) => {
    const [cmd, parseFn] = list(options)
    return parseFn(await spawnProgress(cmd, {}))
  },
  is: (packageManager: "yarn" | "npm"): boolean => (packageManager === "yarn" ? yarn() : !yarn()),
}
