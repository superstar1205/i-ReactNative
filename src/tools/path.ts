import { GluegunFilesystem } from "gluegun"
import * as path from "path"

export const BOILERPLATE_PATH = path.join(__dirname, "../..", "/boilerplate")

export function createGetAllFilePaths(filesystem: GluegunFilesystem) {
  /**
   * Recursively get all file paths in a directory
   */
  return function getAllFilePaths(dir: string): string[] {
    const files = filesystem.list(dir)
    const filePaths = files
      .map((file) => `${dir}/${file}`)
      .filter((file) => filesystem.isFile(file))
    const subdirs = files.filter((file) => filesystem.isDirectory(`${dir}/${file}`))
    const subdirPaths = subdirs.map((subdir) => getAllFilePaths(`${dir}/${subdir}`))
    return [...filePaths, ...subdirPaths.flat()]
  }
}
