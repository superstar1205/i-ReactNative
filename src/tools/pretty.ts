import { print } from "gluegun/print"

const { cyan, gray, white, bold, red, yellow } = print.colors
const { underline } = print.colors

export const p = (m = "") => print.info(gray(`   ${m}`))
export const heading = (m = "") => p(white(bold(m)))
export const link = (m = "") => underline(white(m))
// export const igniteHeading = (m = "") => p(red(bold(m)))
export const igniteHeading = () =>
  p(
    red(
      bold(
        "· · · · · · · · · · · · · · · · · · 🔥 Ignite 🔥 · · · · · · · · · · · · · · · · · ·\n",
      ),
    ),
  )
export const command = (m = "", second = "", examples: string[] = []) => {
  p(white(m) + "  " + gray(second))
  const indent = m.length + 2
  if (examples) {
    examples.forEach((ex) => p(gray(" ".repeat(indent) + ex)))
  }
}
export const direction = (m = "") => p(cyan(m))
export const warning = (m = "") => p(yellow(m))
