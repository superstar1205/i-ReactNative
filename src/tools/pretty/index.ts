export const INDENT = "   "

export * from "./print"
export * from "./prompt"
export * from "./spinners"
export * from "./theme"
