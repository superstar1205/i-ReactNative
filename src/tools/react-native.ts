import { GluegunToolbox } from "gluegun"
import { children } from "./filesystem-ext"

export const isAndroidInstalled = (toolbox: GluegunToolbox): boolean => {
  const androidHome = process.env.ANDROID_HOME
  const hasAndroidEnv = !toolbox.strings.isBlank(androidHome)
  const hasAndroid = hasAndroidEnv && toolbox.filesystem.exists(`${androidHome}/tools`) === "dir"

  return Boolean(hasAndroid)
}

type CopyBoilerplateOptions = {
  boilerplatePath: string
  projectName: string
  excluded: Array<string>
}

/**
 * Copies the boilerplate over to the destination folder.
 *
 */
export async function copyBoilerplate(toolbox: GluegunToolbox, options: CopyBoilerplateOptions) {
  const { filesystem } = toolbox
  const { copyAsync, path } = filesystem

  // ensure the destination folder exists
  await filesystem.dirAsync(options.projectName)

  // rather than copying everything wholesale, let's check what's in the boilerplate folder
  // and copy over everything except stuff like lockfiles and node_modules
  // just to make it faster, y'know? Don't want to copy unnecessary stuff
  const filesAndFolders = children(options.boilerplatePath, true)
  const copyTargets = filesAndFolders.filter(
    (file) => !options.excluded.find((exclusion) => file.includes(exclusion)),
  )

  // kick off a bunch of copies
  const copyPromises = copyTargets.map((fileOrFolder) =>
    copyAsync(path(options.boilerplatePath, fileOrFolder), path(options.projectName, fileOrFolder)),
  )

  // copy them all at once
  return Promise.all(copyPromises)
}

export async function renameReactNativeApp(
  toolbox: GluegunToolbox,
  oldName: string,
  newName: string,
  oldBundleIdentifier: string,
  newBundleIdentifier: string,
) {
  const { parameters, filesystem, print, strings } = toolbox
  const { kebabCase } = strings
  const { path } = filesystem

  // debug?
  const debug = Boolean(parameters.options.debug)
  const log = <T = unknown>(m: T): T => {
    debug && print.info(` ${m}`)
    return m
  }

  // lower case stuff
  const oldnamelower = oldName.toLowerCase()
  const newnamelower = newName.toLowerCase()

  // kebab case
  const oldnamekebab = kebabCase(oldName)
  const newnamekebab = kebabCase(newName)

  // SCREAMING_SNAKE_CASE
  const oldnamesnake = oldnamelower.replace(/[^a-z0-9]/g, "_").toUpperCase()
  const newnamesnake = newnamelower.replace(/[^a-z0-9]/g, "_").toUpperCase()

  async function rename(oldFile: string, newFile: string) {
    log(`Renaming ${oldFile} to ${newFile}`)
    return filesystem.renameAsync(oldFile, newFile)
  }

  // rename files and folders

  // prettier-ignore
  await Promise.allSettled([
    rename(`android/app/src/main/java/com/${oldnamelower}`, newnamelower),
    rename(`android/app/src/debug/java/com/${oldnamelower}`, newnamelower),
    rename(`ios/${oldName}.xcodeproj/xcshareddata/xcschemes/${oldName}.xcscheme`, `${newName}.xcscheme`),
    rename(`ios/${oldName}Tests/${oldName}Tests.m`, `${newName}Tests.m`),
    rename(`ios/${oldName}-Bridging-Header.h`, `${newName}-Bridging-Header.h`),
    rename(`ios/${oldName}.xcworkspace`, `${newName}.xcworkspace`),
    rename(`ios/${oldName}`, `${newName}`),
  ])

  // these we delay to avoid race conditions
  await Promise.allSettled([
    rename(`ios/${oldName}Tests`, `${newName}Tests`),
    rename(`ios/${oldName}.xcodeproj`, `${newName}.xcodeproj`),
  ])

  // if the bundle identifier / android package name changed,
  // we need to move everything to the new folder structure
  const oldPath = oldBundleIdentifier.replace(/\./g, "/")
  const newPath = newBundleIdentifier.replace(/\./g, "/")

  if (oldBundleIdentifier !== newBundleIdentifier) {
    log(`Renaming bundle identifier to ${newBundleIdentifier}`)

    // move everything at the old bundle identifier path to the new one
    await Promise.allSettled([
      moveDeepFolder(
        toolbox,
        `android/app/src/main/java/${oldPath}`,
        `android/app/src/main/java/${newPath}`,
      ),
      moveDeepFolder(
        toolbox,
        `android/app/src/debug/java/${oldPath}`,
        `android/app/src/debug/java/${newPath}`,
      ),
    ])
  }

  // here's a list of all the files to patch the name in
  const filesToPatch = [
    `app.json`,
    `package.json`,
    `index.js`,
    `android/settings.gradle`,
    `android/app/_BUCK`,
    `android/app/BUCK`,
    `android/app/build.gradle`,
    `android/app/src/debug/java/${newPath}/ReactNativeFlipper.java`,
    `android/app/src/main/AndroidManifest.xml`,
    `android/app/src/main/java/${newPath}/MainActivity.java`,
    `android/app/src/main/java/${newPath}/MainApplication.java`,
    `android/app/src/main/java/${newPath}/MainApplication.java`,
    `android/app/src/main/java/${newPath}/newarchitecture/MainApplicationReactNativeHost.java`,
    `android/app/src/main/java/${newPath}/newarchitecture/components/MainComponentsRegistry.java`,
    `android/app/src/main/java/${newPath}/newarchitecture/modules/MainApplicationTurboModuleManagerDelegate.java`,
    `android/app/src/main/jni/Android.mk`,
    `android/app/src/main/jni/MainApplicationTurboModuleManagerDelegate.h`,
    `android/app/src/main/jni/MainComponentsRegistry.h`,
    `android/app/src/main/res/values/strings.xml`,
    `ios/Podfile`,
    `ios/${newName}/Info.plist`,
    `ios/${newName}.xcodeproj/project.pbxproj`,
    `ios/${newName}.xcodeproj/xcshareddata/xcschemes/${newName}.xcscheme`,
    `ios/${newName}.xcworkspace/contents.xcworkspacedata`,
    `ios/${newName}Tests/${newName}Tests.m`,
    `ios/${newName}/AppDelegate.mm`,
    `ios/${newName}/LaunchScreen.storyboard`,
  ]

  // patch the files
  await Promise.allSettled(
    filesToPatch.map(async (file) => {
      // no need to patch files that don't exist
      const exists = await filesystem.existsAsync(path(file))
      if (!exists) return

      const content = await filesystem.readAsync(path(process.cwd(), file), "utf8")

      log(`Patching ${file} - ${oldName} to ${newName} and variants`)

      // replace all instances of the old name and all its variants
      const newContent = content
        .replace(new RegExp(oldBundleIdentifier, "g"), newBundleIdentifier)
        .replace(new RegExp(oldnamekebab, "g"), newnamekebab)
        .replace(new RegExp(oldnamesnake, "g"), newnamesnake)
        .replace(new RegExp(oldName, "g"), newName)
        .replace(new RegExp(oldnamelower, "g"), newnamelower)

      // write the new content back to the file
      await filesystem.writeAsync(file, newContent, { atomic: true })
    }),
  )
}

async function moveDeepFolder(toolbox: GluegunToolbox, oldPath: string, newPath: string) {
  const { filesystem } = toolbox
  const { path } = filesystem

  // ensure newPath exists
  await filesystem.dirAsync(newPath)

  // move the files
  const filesToMove = children(oldPath)
  const movePromises = filesToMove.map((file) =>
    filesystem.moveAsync(path(oldPath, file), path(newPath, file)),
  )

  await Promise.allSettled(movePromises)

  // remove the old folder
  await filesystem.removeAsync(oldPath)
}
