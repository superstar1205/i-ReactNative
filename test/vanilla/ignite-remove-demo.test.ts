import { filesystem } from "gluegun"
import * as tempy from "tempy"

import { runIgnite } from "../_test-helpers"
import { BOILERPLATE_PATH } from "../../src/tools/path"

const setup = (): { TEMP_DIR: string } => {
  const TEMP_DIR = tempy.directory({ prefix: "ignite-" })

  beforeEach(() => {
    filesystem.copy(BOILERPLATE_PATH, TEMP_DIR, { overwrite: true })
  })

  afterEach(() => {
    filesystem.remove(TEMP_DIR) // clean up our mess
  })

  return { TEMP_DIR }
}

describe("ignite-cli remove-demo", () => {
  const { TEMP_DIR } = setup()

  it("should print the expected response", async () => {
    const result = await runIgnite(`remove-demo ${TEMP_DIR}`)

    const MOCK_DIR = `/user/home/ignite`
    const output = result.replace(new RegExp(TEMP_DIR, "g"), MOCK_DIR)
    expect(output).toMatchSnapshot()
  })
})
