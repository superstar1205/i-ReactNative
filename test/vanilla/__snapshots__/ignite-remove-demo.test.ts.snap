// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ignite-cli remove-demo should print the expected response 1`] = `
"   
   Removing demo code from '/user/home/ignite'
   Found '@demo remove-current-line', '@demo remove-block-start', '@demo remove-block-end' in /user/home/ignite/app/i18n/ar.ts
   Found '@demo remove-current-line', '@demo remove-block-start', '@demo remove-block-end' in /user/home/ignite/app/i18n/en.ts
   Found '@demo remove-current-line', '@demo remove-block-start', '@demo remove-block-end' in /user/home/ignite/app/i18n/ko.ts
   Found '@demo remove-file' in /user/home/ignite/app/models/AuthenticationStore.ts
   Found '@demo remove-file' in /user/home/ignite/app/models/Episode.test.ts
   Found '@demo remove-file' in /user/home/ignite/app/models/Episode.ts
   Found '@demo remove-file' in /user/home/ignite/app/models/EpisodeStore.ts
   Found '@demo remove-current-line' in /user/home/ignite/app/models/RootStore.ts
   Found '@demo remove-current-line', '@demo remove-block-start', '@demo remove-block-end' in /user/home/ignite/app/navigators/AppNavigator.tsx
   Found '@demo remove-file' in /user/home/ignite/app/navigators/DemoNavigator.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoCommunityScreen.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoComponentsScreen/DemoComponentsScreen.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoComponentsScreen/DemoDivider.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoComponentsScreen/demos/DemoAutoImage.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoComponentsScreen/demos/DemoButton.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoComponentsScreen/demos/DemoCard.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoComponentsScreen/demos/DemoHeader.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoComponentsScreen/demos/DemoIcon.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoComponentsScreen/demos/DemoListItem.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoComponentsScreen/demos/DemoText.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoComponentsScreen/demos/DemoTextField.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoComponentsScreen/demos/DemoToggle.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoComponentsScreen/demos/index.ts
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoComponentsScreen/DemoUseCase.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoComponentsScreen/DrawerIconButton.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoDebugScreen.tsx
   Found '@demo remove-file' in /user/home/ignite/app/screens/DemoPodcastListScreen.tsx
   Found '@demo remove-block-start', '@demo remove-block-end' in /user/home/ignite/app/screens/index.ts
   Found '@demo remove-file' in /user/home/ignite/app/screens/LoginScreen.tsx
   Found '@demo remove-current-line', '@demo remove-block-start', '@demo remove-block-end' in /user/home/ignite/app/screens/WelcomeScreen.tsx
   Found '@demo remove-current-line', '@demo remove-block-start', '@demo remove-block-end' in /user/home/ignite/app/services/api/api.ts
   Removed empty directory '/user/home/ignite/app/screens/DemoComponentsScreen/demos'
   Done removing demo code from '/user/home/ignite'
"
`;
